File: bindings/wasm/iota_interaction_ts/lib/move_calls/identity/create.ts
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { bcs } from "@iota/iota-sdk/bcs";
import { Transaction } from "@iota/iota-sdk/transactions";
import { getClockRef } from "../utils";

export async function create(didDoc: Uint8Array | undefined, packageId: string): Promise<Uint8Array> {
    const tx = new Transaction();
    const didDocArg = tx.pure(bcs.option(bcs.vector(bcs.U8)).serialize(didDoc));
    const clock = getClockRef(tx);

    tx.moveCall({
        target: `${packageId}::identity::new`,
        arguments: [didDocArg, clock],
    });

    return tx.build({onlyTransactionKind: true});
}

export async function newWithControllers(
    didDoc: Uint8Array | undefined,
    controllers: [string, number][],
    threshold: number,
    packageId: string,
): Promise<Uint8Array> {
    const tx = new Transaction();
    const ids = tx.pure.vector("address", controllers.map(controller => controller[0]));
    const vps = tx.pure.vector("u64", controllers.map(controller => controller[1]));
    const controllersArg = tx.moveCall({
        target: `${packageId}::utils::vec_map_from_keys_values`,
        typeArguments: ["address", "u64"],
        arguments: [ids, vps],
    });
    const controllersThatCanDelegate = tx.moveCall({
        target: "0x2::vec_map::empty",
        typeArguments: ["address", "u64"],
        arguments: [],
    });
    const didDocArg = tx.pure(bcs.option(bcs.vector(bcs.U8)).serialize(didDoc));
    const clock = getClockRef(tx);
    const thresholdArg = tx.pure.u64(threshold);

    tx.moveCall({
        target: `${packageId}::identity::new_with_controllers`,
        arguments: [didDocArg, controllersArg, controllersThatCanDelegate, thresholdArg, clock],
    });

    const tx_kind_bcs = await tx.build({onlyTransactionKind: true});
    try {
        const tx_kind = bcs.TransactionKind.parse(tx_kind_bcs);
    } catch(e) {
        console.error(`failed to deserialize tx kind: ${e}`);
    } finally {
        return tx_kind_bcs;
    }
}

File: identity_iota_core/src/rebased/iota/move_calls/identity/create.rs
// Copyright 2020-2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use iota_interaction::types::base_types::IotaAddress;
use iota_interaction::types::base_types::ObjectID;
use iota_interaction::types::programmable_transaction_builder::ProgrammableTransactionBuilder;
use iota_interaction::types::transaction::ProgrammableTransaction;
use iota_interaction::types::TypeTag;
use iota_interaction::types::IOTA_FRAMEWORK_PACKAGE_ID;
use iota_interaction::ident_str;

use crate::rebased::iota::move_calls::utils;
use crate::rebased::Error;

/// Build a transaction that creates a new on-chain Identity containing `did_doc`.
pub(crate) fn new(did_doc: &[u8], package_id: ObjectID) -> Result<ProgrammableTransaction, Error> {
  let mut ptb = ProgrammableTransactionBuilder::new();
  let doc_arg = utils::ptb_pure(&mut ptb, "did_doc", did_doc)?;
  let clock = utils::get_clock_ref(&mut ptb);

  // Create a new identity, sending its capability to the tx's sender.
  let _identity_id = ptb.programmable_move_call(
    package_id,
    ident_str!("identity").into(),
    ident_str!("new").into(),
    vec![],
    vec![doc_arg, clock],
  );

  Ok(ptb.finish())
}

pub(crate) fn new_with_controllers<C>(
  did_doc: &[u8],
  controllers: C,
  threshold: u64,
  package_id: ObjectID,
) -> Result<ProgrammableTransaction, Error>
where
  C: IntoIterator<Item = (IotaAddress, u64)>,
{
  let mut ptb = ProgrammableTransactionBuilder::new();

  let controllers = {
    let (ids, vps): (Vec<IotaAddress>, Vec<u64>) = controllers.into_iter().unzip();
    let ids = ptb.pure(ids).map_err(|e| Error::InvalidArgument(e.to_string()))?;
    let vps = ptb.pure(vps).map_err(|e| Error::InvalidArgument(e.to_string()))?;
    ptb.programmable_move_call(
      package_id,
      ident_str!("utils").into(),
      ident_str!("vec_map_from_keys_values").into(),
      vec![TypeTag::Address, TypeTag::U64],
      vec![ids, vps],
    )
  };

  let controllers_that_can_delegate = ptb.programmable_move_call(
    IOTA_FRAMEWORK_PACKAGE_ID,
    ident_str!("vec_map").into(),
    ident_str!("empty").into(),
    vec![TypeTag::Address, TypeTag::U64],
    vec![],
  );
  let doc_arg = ptb.pure(did_doc).map_err(|e| Error::InvalidArgument(e.to_string()))?;
  let threshold_arg = ptb.pure(threshold).map_err(|e| Error::InvalidArgument(e.to_string()))?;
  let clock = utils::get_clock_ref(&mut ptb);

  // Create a new identity, sending its capabilities to the specified controllers.
  let _identity_id = ptb.programmable_move_call(
    package_id,
    ident_str!("identity").into(),
    ident_str!("new_with_controllers").into(),
    vec![],
    vec![
      doc_arg,
      controllers,
      controllers_that_can_delegate,
      threshold_arg,
      clock,
    ],
  );

  Ok(ptb.finish())
}
